File: feature/src/main/java/com/ortin/gesturetranslator/feature/managers/word_compiler/WordCompilerManagerImpl.java
package com.ortin.gesturetranslator.feature.managers.word_compiler;

import java.util.HashMap;

public class WordCompilerManagerImpl implements WordCompilerManager {

    private final int MIN_FRAME_DETECT = 10;
    private final long MIN_SPACE_TIME = 3000L;

    private String text = "";
    private String lastWord;
    private String currentLastWord;
    private int countFrameHold;
    private long timeSpace;
    private final HashMap<String, Integer> frameSortedMap = new HashMap<>();

    private long lastTime;

    @Override
    public void addLetter(String letter) {
        timeSpace = lastTime != 0L ? System.currentTimeMillis() - lastTime : 0L;

        if (timeSpace >= MIN_SPACE_TIME) text += " ";

        if (lastWord != null) {
            if (lastWord.equals(letter)) {
                countFrameHold++;
                if (countFrameHold >= MIN_FRAME_DETECT){
                    if (currentLastWord != null && !currentLastWord.equals(lastWord))
                        text += lastWord;
                    else if (currentLastWord == null) text += lastWord;
                    currentLastWord = lastWord;
                    frameSortedMap.clear();
                    lastWord = null;
                    countFrameHold = 0;
                }
            } else {
                if (frameSortedMap.containsKey(lastWord))
                    frameSortedMap.put(lastWord, frameSortedMap.get(lastWord) + countFrameHold);
                else frameSortedMap.put(lastWord, countFrameHold);

                countFrameHold = 1;
                lastWord = letter;

                String checkWord = searchCurrentWord();
                if (checkWord != null) {
                    if (currentLastWord != null && !currentLastWord.equals(checkWord))
                        text += checkWord;
                    else if (currentLastWord == null) text += checkWord;
                    currentLastWord = checkWord;
                    frameSortedMap.clear();
                    lastWord = null;
                }
            }
        } else {
            lastWord = letter;
            countFrameHold = 1;
        }


        lastTime = System.currentTimeMillis();
    }

    @Override
    public String getWord() {
        return text;
    }

    private String searchCurrentWord() {
        for (String key : frameSortedMap.keySet()) {
            if (frameSortedMap.get(key) >= MIN_FRAME_DETECT) {
                return key;
            }
        }
        return null;
    }

    @Override
    public void clearState(){
        text = "";
        lastWord = null;
        currentLastWord = null;
        countFrameHold = 0;
        timeSpace = 0L;
        frameSortedMap.clear();
        lastTime = 0L;
    }
}

File: feature/src/main/java/com/ortin/gesturetranslator/feature/managers/word_compiler/WordCompilerManagerImpl.kt
package com.ortin.gesturetranslator.feature.managers.word_compiler

class WordCompilerManagerImpl : WordCompilerManager {
    private val MIN_FRAME_DETECT = 10
    private val MIN_SPACE_TIME = 3000L
    private var text = ""

    private var lastWord: String? = null
    private var currentLastWord: String? = null
    private var countFrameHold = 0

    private var timeSpace: Long = 0
    private val frameSortedMap = HashMap<String, Int>()
    private var lastTime: Long = 0
    override fun addLetter(letter: String?) {
        timeSpace = if (lastTime != 0L) System.currentTimeMillis() - lastTime else 0L
        if (timeSpace >= MIN_SPACE_TIME) text += " "

        if (lastWord != null) {
            if (lastWord == letter) {
                countFrameHold++
                if (countFrameHold >= MIN_FRAME_DETECT) {
                    if (currentLastWord != null && currentLastWord != lastWord) text += lastWord else if (currentLastWord == null) text += lastWord
                    currentLastWord = lastWord
                    frameSortedMap.clear()
                    lastWord = null
                    countFrameHold = 0
                }
            } else {
                if (frameSortedMap.containsKey(lastWord)) frameSortedMap[lastWord!!] =
                    frameSortedMap[lastWord]!! + countFrameHold else frameSortedMap[lastWord!!] =
                    countFrameHold

                countFrameHold = 1
                lastWord = letter

                val checkWord = searchCurrentWord()

                if (checkWord != null) {
                    if (currentLastWord != null && currentLastWord != checkWord) text += checkWord else if (currentLastWord == null) text += checkWord
                    currentLastWord = checkWord
                    frameSortedMap.clear()
                    lastWord = null
                }
            }
        } else {
            lastWord = letter
            countFrameHold = 1
        }

        lastTime = System.currentTimeMillis()
    }

    override fun getWord(): String {
        return text
    }

    private fun searchCurrentWord(): String? {
        for (key in frameSortedMap.keys) {
            if (frameSortedMap[key]!! >= MIN_FRAME_DETECT) {
                return key
            }
        }
        return null
    }

    override fun clearState() {
        text = ""
        lastWord = null
        currentLastWord = null
        countFrameHold = 0
        timeSpace = 0L
        frameSortedMap.clear()
        lastTime = 0L
    }
}